From e1443019da5507fad8cc398bcf964cf3c72643ab Mon Sep 17 00:00:00 2001
From: Michael Stahl <mstahl@redhat.com>
Date: Mon, 24 Sep 2012 11:21:11 +0200
Subject: [PATCH 3/4] rhbz#820283: fix crashes in DOCX table import:

It is apparently possible that writerfilter creates a table with
irregular structure, where the lines have varying numbers of boxes in
them.  This may cause crashes on later editing operations, e.g.
when SwTable::NewInsertCol() is unable to find boxes covering a
particular column.  To prevent this, add missing boxes in
SwNodes::TextToTable().

(cherry picked from commit 6d2e09db4a677068095b0bebd08fbbb96620d60c)

Change-Id: I423821645baeaf55595e4933e1ab8fb89b2099f3
---
 sw/source/core/docnode/ndtbl.cxx | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/sw/source/core/docnode/ndtbl.cxx b/sw/source/core/docnode/ndtbl.cxx
index ca006bc..03409ce 100644
--- a/sw/source/core/docnode/ndtbl.cxx
+++ b/sw/source/core/docnode/ndtbl.cxx
@@ -1374,6 +1374,20 @@ SwTableNode* SwNodes::TextToTable( const SwNodes::TableRanges_t & rTableNodes,
     // die Tabelle ausgleichen, leere Sections einfuegen
     sal_uInt16 n;
 
+    for (n = 0; n < pTable->GetTabLines().Count(); ++n )
+    {
+        // rhbz#820283: balance the cells in table rows
+        SwTableLine *const pCurrLine = pTable->GetTabLines()[ n ];
+        nBoxes = pCurrLine->GetTabBoxes().Count();
+        SwTxtFmtColl *const pTxtColl(const_cast<SwTxtFmtColl*>(
+                GetDoc()->GetDfltTxtFmtColl()));
+        if (nMaxBoxes != nBoxes)
+        {
+            InsBoxen( pTblNd, pCurrLine, pBoxFmt, pTxtColl, 0,
+                        nBoxes, nMaxBoxes - nBoxes );
+        }
+    }
+
     if( !aPosArr.empty() )
     {
         SwTableLines& rLns = pTable->GetTabLines();
-- 
1.7.11.4

